### PR TITLE
Fix duplicate multibindings error

### PR DESCRIPTION
**Pull request by tcmulcahy**
- Add a minimal repro regression test for a duplicate multi-binding error we experienced internally at Square
- Fix the error by treating repeated discovery of the same multibinding contribution as idempotent
- remove the duplicate multibinding check - a repeated add does not represent a user-facing duplicate binding conflict; it represents repeated discovery of the same contribution

The code and repro are entirely AI generated, but I have carefully reviewed it. I have also verified that this fixes the problem that we were hitting internally (now we are hitting a different problem, which I will fix next).

### CODE REVIEW

File: compiler/src/main/kotlin/dev/zacsweers/metro/compiler/ir/graph/BindingGraphGenerator.kt [hunk 1]
```diff
@@ -634,8 +634,11 @@ internal class BindingGraphGenerator(
             if (key == node.metroGraph?.generatedGraphExtensionData?.typeKey) continue
             // Skip if there's a dynamic replacement for this key
             if (key in node.dynamicTypeKeys) continue
-            val existingBinding = graph.findBinding(key)
-            if (existingBinding != null) {
+
+            // Use bindingLookup as the source of truth. graph.findBinding() only reflects keys
+            // added through graph.addBinding(), which is disabled when full graph validation is
+            // off.
+            if (key in bindingLookup) {
```
Comment (ZacSweers) [inline compiler/src/main/kotlin/dev/zacsweers/metro/compiler/ir/graph/BindingGraphGenerator.kt:639]: Do we need the BindingLookup changes with this fix? Seems like this is the fix and those might just be defensive

Comment (tcmulcahy) [reply compiler/src/main/kotlin/dev/zacsweers/metro/compiler/ir/graph/BindingGraphGenerator.kt:639]: Maybe! I'll give it a try.

Comment (ZacSweers) [reply compiler/src/main/kotlin/dev/zacsweers/metro/compiler/ir/graph/BindingGraphGenerator.kt:639]: Looks like it worked. Wanna add a line to the changelog and we'll get this merged?

Comment (tcmulcahy) [reply compiler/src/main/kotlin/dev/zacsweers/metro/compiler/ir/graph/BindingGraphGenerator.kt:639]: The BindingLookup.kt changes are required. Here's what I get if I revert BindingLookup.kt back to the version in main:

```
org.jetbrains.kotlin.fir.pipeline.IrGenerationExtensionException: Duplicate multibinding source: @dev.zacsweers.metro.internal.MultibindingElement("PreloadService", "-963646100") PreloadService. This is possibly a bug in the Metro compiler, please report it with details and/or a reproducer to https://github.com/zacsweers/metro. 
	at org.jetbrains.kotlin.fir.pipeline.Fir2IrPipeline.applyIrGenerationExtensions(convertToIr.kt:577)
	at org.jetbrains.kotlin.fir.pipeline.Fir2IrPipeline.runActualizationPipeline(convertToIr.kt:279)
	at org.jetbrains.kotlin.fir.pipeline.Fir2IrPipeline.convertToIrAndActualize(convertToIr.kt:150)
	at org.jetbrains.kotlin.fir.pipeline.ConvertToIrKt.convertToIrAndActualize(convertToIr.kt:115)
	at org.jetbrains.kotlin.fir.pipeline.ConvertToIrKt.convertToIrAndActualize$default(convertToIr.kt:88)
	at org.jetbrains.kotlin.cli.jvm.compiler.legacy.pipeline.JvmCompilerPipelineKt.convertToIrAndActualizeForJvm(jvmCompilerPipeline.kt:100)
	at org.jetbrains.kotlin.cli.pipeline.jvm.JvmFir2IrPipelinePhase.executePhase(JvmFir2IrPipelinePhase.kt:30)
	at org.jetbrains.kotlin.cli.pipeline.jvm.JvmFir2IrPipelinePhase.executePhase(JvmFir2IrPipelinePhase.kt:25)
	at org.jetbrains.kotlin.cli.pipeline.jvm.JvmFir2IrPipelinePhase.executePhase(JvmFir2IrPipelinePhase.kt:19)
	at org.jetbrains.kotlin.test.frontend.fir.Fir2IrCliFacade.transform(Fir2IrCliFacade.kt:49)
	at org.jetbrains.kotlin.test.frontend.fir.Fir2IrCliFacade.transform(Fir2IrCliFacade.kt:24)
	at org.jetbrains.kotlin.test.TestStep$FacadeStep.processModule(TestStep.kt:47)
	at org.jetbrains.kotlin.test.TestRunner.processModule(TestRunner.kt:206)
	at org.jetbrains.kotlin.test.TestRunner.hackyProcessModule(TestRunner.kt:197)
	at org.jetbrains.kotlin.test.TestRunner.processModule(TestRunner.kt:141)
	at org.jetbrains.kotlin.test.TestRunner.runTestPipeline(TestRunner.kt:92)
	at org.jetbrains.kotlin.test.TestRunner.runTestImpl(TestRunner.kt:70)
	at org.jetbrains.kotlin.test.TestRunner.runTest(TestRunner.kt:30)
	at org.jetbrains.kotlin.test.TestRunner.runTest$default(TestRunner.kt:28)
	at org.jetbrains.kotlin.test.runners.AbstractKotlinCompilerTest.runTest(AbstractKotlinCompilerTest.kt:115)
	at dev.zacsweers.metro.compiler.BoxTestGenerated$Interop$Dagger$Anvil.testDuplicateMultibindingSourcesFromSharedIncludedModule(BoxTestGenerated.java:2018)
	at java.base/java.lang.reflect.Method.invoke(Method.java:565)
	at java.base/java.util.concurrent.RecursiveAction.exec(RecursiveAction.java:194)
	at java.base/java.util.concurrent.ForkJoinTask.doExec(ForkJoinTask.java:507)
	at java.base/java.util.concurrent.ForkJoinPool$WorkQueue.topLevelExec(ForkJoinPool.java:1394)
	at java.base/java.util.concurrent.ForkJoinPool.runWorker(ForkJoinPool.java:1970)
	at java.base/java.util.concurrent.ForkJoinWorkerThread.run(ForkJoinWorkerThread.java:187)
Caused by: java.lang.IllegalStateException: Duplicate multibinding source: @dev.zacsweers.metro.internal.MultibindingElement("PreloadService", "-963646100") PreloadService. This is possibly a bug in the Metro compiler, please report it with details and/or a reproducer to https://github.com/zacsweers/metro. 
	at dev.zacsweers.metro.compiler.ir.graph.IrBinding$Multibinding.addSourceBinding(IrBinding.kt:981)
	at dev.zacsweers.metro.compiler.ir.graph.BindingLookup.registerMultibindingContribution(BindingLookup.kt:307)
	at dev.zacsweers.metro.compiler.ir.graph.BindingLookup.putBinding(BindingLookup.kt:190)
	at dev.zacsweers.metro.compiler.ir.graph.BindingGraphGenerator.generate$putBinding(BindingGraphGenerator.kt:97)
	at dev.zacsweers.metro.compiler.ir.graph.BindingGraphGenerator.access$generate$putBinding(BindingGraphGenerator.kt:46)
	at dev.zacsweers.metro.compiler.ir.graph.BindingGraphGenerator.generate(BindingGraphGenerator.kt:312)
	at dev.zacsweers.metro.compiler.ir.transformers.DependencyGraphTransformer.validateDependencyGraph(DependencyGraphTransformer.kt:242)
	at dev.zacsweers.metro.compiler.ir.transformers.DependencyGraphTransformer.access$validateDependencyGraph(DependencyGraphTransformer.kt:106)
	at dev.zacsweers.metro.compiler.ir.transformers.DependencyGraphTransformer.validateDependencyGraph$validateExtension(DependencyGraphTransformer.kt:356)
	at dev.zacsweers.metro.compiler.ir.transformers.DependencyGraphTransformer.validateDependencyGraph(DependencyGraphTransformer.kt:394)
	at dev.zacsweers.metro.compiler.ir.transformers.DependencyGraphTransformer.access$validateDependencyGraph(DependencyGraphTransformer.kt:106)
	at dev.zacsweers.metro.compiler.ir.transformers.DependencyGraphTransformer.validateDependencyGraph$validateExtension(DependencyGraphTransformer.kt:356)
	at dev.zacsweers.metro.compiler.ir.transformers.DependencyGraphTransformer.validateDependencyGraph(DependencyGraphTransformer.kt:394)
	at dev.zacsweers.metro.compiler.ir.transformers.DependencyGraphTransformer.access$validateDependencyGraph(DependencyGraphTransformer.kt:106)
	at dev.zacsweers.metro.compiler.ir.transformers.DependencyGraphTransformer.processGraphInner(DependencyGraphTransformer.kt:170)
	at dev.zacsweers.metro.compiler.ir.transformers.DependencyGraphTransformer.processGraph$compiler(DependencyGraphTransformer.kt:130)
	at dev.zacsweers.metro.compiler.ir.MetroIrGenerationExtension.generateInner(MetroIrGenerationExtension.kt:189)
	at dev.zacsweers.metro.compiler.ir.MetroIrGenerationExtension.generate(MetroIrGenerationExtension.kt:70)
	at org.jetbrains.kotlin.fir.pipeline.Fir2IrPipeline.applyIrGenerationExtensions(convertToIr.kt:571)
	... 26 more
Cause 1: java.lang.IllegalStateException: Duplicate multibinding source: @dev.zacsweers.metro.internal.MultibindingElement("PreloadService", "-963646100") PreloadService. This is possibly a bug in the Metro compiler, please report it with details and/or a reproducer to https://github.com/zacsweers/metro. 
	at dev.zacsweers.metro.compiler.ir.graph.IrBinding$Multibinding.addSourceBinding(IrBinding.kt:981)
	at dev.zacsweers.metro.compiler.ir.graph.BindingLookup.registerMultibindingContribution(BindingLookup.kt:307)
	at dev.zacsweers.metro.compiler.ir.graph.BindingLookup.putBinding(BindingLookup.kt:190)
	at dev.zacsweers.metro.compiler.ir.graph.BindingGraphGenerator.generate$putBinding(BindingGraphGenerator.kt:97)
	at dev.zacsweers.metro.compiler.ir.graph.BindingGraphGenerator.access$generate$putBinding(BindingGraphGenerator.kt:46)
	at dev.zacsweers.metro.compiler.ir.graph.BindingGraphGenerator.generate(BindingGraphGenerator.kt:312)
	at dev.zacsweers.metro.compiler.ir.transformers.DependencyGraphTransformer.validateDependencyGraph(DependencyGraphTransformer.kt:242)
	at dev.zacsweers.metro.compiler.ir.transformers.DependencyGraphTransformer.access$validateDependencyGraph(DependencyGraphTransformer.kt:106)
	at dev.zacsweers.metro.compiler.ir.transformers.DependencyGraphTransformer.validateDependencyGraph$validateExtension(DependencyGraphTransformer.kt:356)
	at dev.zacsweers.metro.compiler.ir.transformers.DependencyGraphTransformer.validateDependencyGraph(DependencyGraphTransformer.kt:394)
	at dev.zacsweers.metro.compiler.ir.transformers.DependencyGraphTransformer.access$validateDependencyGraph(DependencyGraphTransformer.kt:106)
	at dev.zacsweers.metro.compiler.ir.transformers.DependencyGraphTransformer.validateDependencyGraph$validateExtension(DependencyGraphTransformer.kt:356)
	at dev.zacsweers.metro.compiler.ir.transformers.DependencyGraphTransformer.validateDependencyGraph(DependencyGraphTransformer.kt:394)
	at dev.zacsweers.metro.compiler.ir.transformers.DependencyGraphTransformer.access$validateDependencyGraph(DependencyGraphTransformer.kt:106)
	at dev.zacsweers.metro.compiler.ir.transformers.DependencyGraphTransformer.processGraphInner(DependencyGraphTransformer.kt:170)
	at dev.zacsweers.metro.compiler.ir.transformers.DependencyGraphTransformer.processGraph$compiler(DependencyGraphTransformer.kt:130)
	at dev.zacsweers.metro.compiler.ir.MetroIrGenerationExtension.generateInner(MetroIrGenerationExtension.kt:189)
	at dev.zacsweers.metro.compiler.ir.MetroIrGenerationExtension.generate(MetroIrGenerationExtension.kt:70)
	at org.jetbrains.kotlin.fir.pipeline.Fir2IrPipeline.applyIrGenerationExtensions(convertToIr.kt:571)
	at org.jetbrains.kotlin.fir.pipeline.Fir2IrPipeline.runActualizationPipeline(convertToIr.kt:279)
	at org.jetbrains.kotlin.fir.pipeline.Fir2IrPipeline.convertToIrAndActualize(convertToIr.kt:150)
	at org.jetbrains.kotlin.fir.pipeline.ConvertToIrKt.convertToIrAndActualize(convertToIr.kt:115)
	at org.jetbrains.kotlin.fir.pipeline.ConvertToIrKt.convertToIrAndActualize$default(convertToIr.kt:88)
	at org.jetbrains.kotlin.cli.jvm.compiler.legacy.pipeline.JvmCompilerPipelineKt.convertToIrAndActualizeForJvm(jvmCompilerPipeline.kt:100)
	at org.jetbrains.kotlin.cli.pipeline.jvm.JvmFir2IrPipelinePhase.executePhase(JvmFir2IrPipelinePhase.kt:30)
	at org.jetbrains.kotlin.cli.pipeline.jvm.JvmFir2IrPipelinePhase.executePhase(JvmFir2IrPipelinePhase.kt:25)
	at org.jetbrains.kotlin.cli.pipeline.jvm.JvmFir2IrPipelinePhase.executePhase(JvmFir2IrPipelinePhase.kt:19)
	at org.jetbrains.kotlin.test.frontend.fir.Fir2IrCliFacade.transform(Fir2IrCliFacade.kt:49)
	at org.jetbrains.kotlin.test.frontend.fir.Fir2IrCliFacade.transform(Fir2IrCliFacade.kt:24)
	at org.jetbrains.kotlin.test.TestStep$FacadeStep.processModule(TestStep.kt:47)
	at org.jetbrains.kotlin.test.TestRunner.processModule(TestRunner.kt:206)
	at org.jetbrains.kotlin.test.TestRunner.hackyProcessModule(TestRunner.kt:197)
	at org.jetbrains.kotlin.test.TestRunner.processModule(TestRunner.kt:141)
	at org.jetbrains.kotlin.test.TestRunner.runTestPipeline(TestRunner.kt:92)
	at org.jetbrains.kotlin.test.TestRunner.runTestImpl(TestRunner.kt:70)
	at org.jetbrains.kotlin.test.TestRunner.runTest(TestRunner.kt:30)
	at org.jetbrains.kotlin.test.TestRunner.runTest$default(TestRunner.kt:28)
	at org.jetbrains.kotlin.test.runners.AbstractKotlinCompilerTest.runTest(AbstractKotlinCompilerTest.kt:115)
	at dev.zacsweers.metro.compiler.BoxTestGenerated$Interop$Dagger$Anvil.testDuplicateMultibindingSourcesFromSharedIncludedModule(BoxTestGenerated.java:2018)
	at java.base/java.lang.reflect.Method.invoke(Method.java:565)
	at java.base/java.util.concurrent.RecursiveAction.exec(RecursiveAction.java:194)
	at java.base/java.util.concurrent.ForkJoinTask.doExec(ForkJoinTask.java:507)
	at java.base/java.util.concurrent.ForkJoinPool$WorkQueue.topLevelExec(ForkJoinPool.java:1394)
	at java.base/java.util.concurrent.ForkJoinPool.runWorker(ForkJoinPool.java:1970)
	at java.base/java.util.concurrent.ForkJoinWorkerThread.run(ForkJoinWorkerThread.java:187)
```

(The refactorization I made introducing addMultibindingContribution is NOT required - that's just a code cleanup and I can undo that if you prefer)

Comment (ZacSweers) [reply compiler/src/main/kotlin/dev/zacsweers/metro/compiler/ir/graph/BindingGraphGenerator.kt:639]: Gotcha. Was able to clean it up a bit higher up by matching existing logic we had for provides. Thanks! [a3dd9df](https://github.com/ZacSweers/metro/pull/1960/commits/a3dd9dfd6fcb8b7eeef7b2fbbf01597b526868c1)

Comment (tcmulcahy) [reply compiler/src/main/kotlin/dev/zacsweers/metro/compiler/ir/graph/BindingGraphGenerator.kt:639]: Looks great. I see you added the CHANGELOG line too. Thanks!
